File: apps/rag-service/app/rag/chunker.py
import hashlib
import re
from dataclasses import dataclass


@dataclass(frozen=True)
class Chunk:
    heading: str
    content: str
    content_hash: str


def _sha256(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _words(text: str) -> list[str]:
    return re.findall(r"\S+", text, flags=re.UNICODE)


def chunk_markdown(markdown: str, *, chunk_size: int = 512,
                   chunk_overlap: int = 64, chunker_version: str = "1") -> list[Chunk]:
    """Heading-scoped chunker (Go indexer behavior) with overlap for oversized bodies."""
    chunks: list[Chunk] = []
    for heading, body in _split_by_headings(markdown):
        body = body.strip()
        if not body:
            continue
        for part in _split_with_overlap(body, chunk_size, chunk_overlap):
            chunks.append(Chunk(heading=heading.strip(), content=part, content_hash=_sha256(part)))
    return chunks


def _split_by_headings(markdown: str):
    """Yield (heading, body) pairs. '# ' title lines are skipped."""
    lines = markdown.replace("\r\n", "\n").split("\n")
    current_heading, current_body, title_seen = "", [], False
    for line in lines:
        if re.match(r"^# ", line) and not title_seen:
            title_seen = True
            continue
        if re.match(r"^#{2,3} ", line):
            if current_body:
                yield current_heading, "\n".join(current_body)
            current_heading = re.sub(r"^#+\s*", "", line)
            current_body = []
        else:
            current_body.append(line)
    if current_body:
        yield current_heading, "\n".join(current_body)


def _split_with_overlap(body: str, chunk_size: int, chunk_overlap: int) -> list[str]:
    """Split body into parts of at most chunk_size words.

    Parts prefer paragraph boundaries; consecutive parts share the last
    chunk_overlap words of the previous part. A paragraph that alone
    exceeds chunk_size is pre-sliced into step-sized units so the
    overlap tail always fits when units are merged.
    """
    if chunk_size < 1:
        raise ValueError("chunk_size must be >= 1")
    if not 0 <= chunk_overlap < chunk_size:
        raise ValueError("chunk_overlap must satisfy 0 <= chunk_overlap < chunk_size")

    words = _words(body)
    if len(words) <= chunk_size:
        return [body]

    step = chunk_size - chunk_overlap  # >= 1 after validation
    units: list[list[str]] = []
    for para in (p.strip() for p in body.split("\n\n") if p.strip()):
        pw = _words(para)
        if len(pw) > chunk_size:
            units.extend(pw[i:i + step] for i in range(0, len(pw), step))
        else:
            units.append(pw)

    parts: list[str] = []
    buffer: list[str] = []
    i = 0
    while i < len(units) or buffer:
        if not buffer:
            buffer = units[i]
            i += 1
            continue
        if i < len(units) and len(buffer) + len(units[i]) <= chunk_size:
            buffer.extend(units[i])
            i += 1
            continue
        # cannot extend further -> flush; next part starts with the overlap tail
        parts.append(" ".join(buffer))
        overlap = buffer[-chunk_overlap:] if chunk_overlap else []
        if i >= len(units):
            buffer = []
        elif len(overlap) + len(units[i]) <= chunk_size:
            buffer = overlap + units[i]
            i += 1
        else:
            fit = chunk_size - len(overlap)
            buffer = overlap + units[i][:fit]
            units[i] = units[i][fit:]  # remainder stays queued
            if not units[i]:
                i += 1
    return [p for p in parts if p]

File: apps/ai-service/internal/knowledge/chunker.go
package knowledge

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
)

type ChunkItem struct {
	Heading     string
	Content     string
	ContentHash string
}

var (
	titleRegex   = regexp.MustCompile(`^#\s+`)
	headingRegex = regexp.MustCompile(`^#{2,3}\s+`)
	stripHRegex  = regexp.MustCompile(`^#+\s*`)
)

func sha256Hash(text string) string {
	h := sha256.Sum256([]byte(text))
	return hex.EncodeToString(h[:])
}

// ChunkMarkdown splits a markdown document by headings (H2, H3) and applies word-based overlap.
func ChunkMarkdown(markdown string, chunkSize, chunkOverlap int, chunkerVersion string) ([]ChunkItem, error) {
	if chunkSize <= 0 {
		return nil, fmt.Errorf("chunk_size must be > 0")
	}
	if chunkOverlap < 0 || chunkOverlap >= chunkSize {
		return nil, fmt.Errorf("chunk_overlap must satisfy 0 <= chunk_overlap < chunk_size")
	}

	sections := splitByHeadings(markdown)
	var chunks []ChunkItem

	for _, sec := range sections {
		body := strings.TrimSpace(sec.body)
		if body == "" {
			continue
		}
		parts := splitWithOverlap(body, chunkSize, chunkOverlap)
		for _, part := range parts {
			chunks = append(chunks, ChunkItem{
				Heading:     strings.TrimSpace(sec.heading),
				Content:     part,
				ContentHash: sha256Hash(part),
			})
		}
	}
	return chunks, nil
}

type headingSection struct {
	heading string
	body    string
}

func splitByHeadings(markdown string) []headingSection {
	raw := strings.ReplaceAll(markdown, "\r\n", "\n")
	lines := strings.Split(raw, "\n")

	var sections []headingSection
	var currentHeading string
	var currentBody []string
	titleSeen := false

	for _, line := range lines {
		if titleRegex.MatchString(line) && !titleSeen {
			titleSeen = true
			continue
		}
		if headingRegex.MatchString(line) {
			if len(currentBody) > 0 {
				sections = append(sections, headingSection{
					heading: currentHeading,
					body:    strings.Join(currentBody, "\n"),
				})
			}
			currentHeading = stripHRegex.ReplaceAllString(line, "")
			currentBody = []string{}
		} else {
			currentBody = append(currentBody, line)
		}
	}

	if len(currentBody) > 0 {
		sections = append(sections, headingSection{
			heading: currentHeading,
			body:    strings.Join(currentBody, "\n"),
		})
	}
	return sections
}

func wordsOf(text string) []string {
	return strings.Fields(text)
}

func splitWithOverlap(body string, chunkSize, chunkOverlap int) []string {
	body = strings.TrimSpace(body)
	if body == "" {
		return nil
	}

	rawParas := strings.Split(body, "\n\n")
	var paras []string
	for _, p := range rawParas {
		trimmed := strings.TrimSpace(p)
		if trimmed != "" {
			paras = append(paras, trimmed)
		}
	}

	// Pre-slice oversized paragraphs
	var units []string
	step := chunkSize - chunkOverlap
	if step <= 0 {
		step = 1
	}

	for _, p := range paras {
		w := wordsOf(p)
		if len(w) <= chunkSize {
			units = append(units, p)
		} else {
			for i := 0; i < len(w); i += step {
				end := i + chunkSize
				if end > len(w) {
					end = len(w)
				}
				units = append(units, strings.Join(w[i:end], " "))
				if end == len(w) {
					break
				}
			}
		}
	}

	var parts []string
	var curWords []string

	for _, unit := range units {
		uWords := wordsOf(unit)
		if len(curWords)+len(uWords) <= chunkSize {
			curWords = append(curWords, uWords...)
		} else {
			if len(curWords) > 0 {
				parts = append(parts, strings.Join(curWords, " "))
				tailStart := len(curWords) - chunkOverlap
				if tailStart < 0 {
					tailStart = 0
				}
				curWords = append([]string{}, curWords[tailStart:]...)
			}
			curWords = append(curWords, uWords...)
			for len(curWords) > chunkSize {
				parts = append(parts, strings.Join(curWords[:chunkSize], " "))
				tailStart := chunkSize - chunkOverlap
				if tailStart < 0 {
					tailStart = 0
				}
				curWords = append([]string{}, curWords[tailStart:]...)
			}
		}
	}

	if len(curWords) > 0 {
		parts = append(parts, strings.Join(curWords, " "))
	}
	return parts
}
